featuturefmt: Extract PotentialNamespace

Allda opened the pull request:
PotentialNamespace is feature namespace extracted while detecting features in a layer. The PotentialNamespace is stored in nemaspace table + reference is also stored in layer_feature.

This will allow having multiple namespaces for single layer feature.

Partially resolves: https://github.com/coreos/clair/issues/706

Allda committed:
featurefmt: Extract PotentialNamespace

PotentialNamespace is feature namespace extracted while detecting
features in layer. It will server for special feature detector. The
current detectors return empty namespace.

Allda committed:
Store PotentialNamespace in database

PotentialNamespace is part of layer_feature table and it is also stored
in namespace table.

KeyboardNerd commented on database/models.go:
```diff
@@ -166,22 +170,23 @@ func NewNamespace(name string, versionFormat string) *Namespace {
+	Version            string      `json:"version"`
+	VersionFormat      string      `json:"versionFormat"`
+	Type               FeatureType `json:"type"`
+	PotentialNamespace Namespace   `json:"potentialNamespace"`
```
Please move PotentialNamespace to LayerFeature struct, because Feature is shared by multiple layers.

Allda replied:
Feature listers return `database.Feature` and not `database.LayerFeature`. Can I change the return type of listers or how can I extract namespace?

KeyboardNerd replied:
yes, please change the feature lister to return LayerFeature.

Allda committed:
Move PotentialNamespace to LayerFeature

PotentialNamespace should be in LayerFeature instead of Feature struct.
Feature extractors were updated to return LayerFeature instead of
Feature.

Allda commented:
@KeyboardNerd The PR is ready to be reviewed. 

Here is the feature lister which uses a potential namespace which I used for testing.
https://gist.github.com/Allda/99d30f5d37d35dd3e2d0244470d6f150

There is one thing which is causing problems now and I believe it will be resolved once you made your changes in updater.go

When feature lister detects multiple namespaces (cpes) it returns each package multiple times with different CPE. It works fine, store namespaces, features, layer_feature, but it fails when storing data to ancestry_feature because of duplicated reference to namespaced_feature_id.

Allda commented:
I found one more bug, wait with a review, please.

Allda committed:
Add namespace_id as constraints to layer_feature

If layer contains more than one potential namespace, features will be
created for each namespace. Layer_feature table now has to contains one
more constrains (namespace_id).

Allda commented:
> I found one more bug, wait with a review, please.

Fixed

KeyboardNerd commented on database/pgsql/migrations/00001_initial_schema.go:
```diff
@@ -89,7 +89,8 @@ var (
 				feature_id INT REFERENCES feature ON DELETE CASCADE,
 				detector_id INT REFERENCES detector ON DELETE CASCADE,
-				UNIQUE (layer_id, feature_id));`,
+				namespace_id INT REFERENCES namespace ON DELETE CASCADE,
```
please explicitly state the namespace_id is nullable

KeyboardNerd commented on database/pgsql/layer.go:
```diff
@@ -37,10 +37,11 @@ const (
+			FROM layer_feature AS lf, feature AS f, feature_type AS t, namespace AS ns
 			WHERE lf.feature_id = f.id
 				AND t.id = f.type
+				AND lf.namespace_id = ns.id
```
what if ls.namespace_id is null? there won't be a matched ns.id.

Allda replied:
Resolved by using Left join

KeyboardNerd commented on database/dbutil.go:
```diff
@@ -58,6 +58,19 @@ func ConvertFeatureSetToFeatures(features mapset.Set) []Feature {
 	return uniqueFeatures
 }
 
+func ConvertFeatureSerToLayerFeatures(features mapset.Set) []LayerFeature {
```
typo

KeyboardNerd commented:
Schema looks ok.

Please add some tests on the change, especially the features with potential namespace.

Allda committed:
Add test for potential namespace

Test verifies that potential namespace is stored in database and it can
be loaded back to structure.

The commit also fixes few typos and bugs.

KeyboardNerd commented on database/pgsql/testdata/data.sql:
```diff
@@ -2,7 +2,8 @@
   (2, 'debian:8', 'dpkg'),
-  (3, 'fake:1.0', 'rpm');
+  (3, 'fake:1.0', 'rpm'),
+  (4, 'cpe:/o:redhat:enterprise_linux:7::server', 'rpm');
```
Does 'cpe:/o:redhat:enterprise_linux:7::server' has any version information?

If this is the actual format for CPE, I'll have to change the namespace db a bit because right now we assume namespace name consists of two components: `<name>:<version>` as a string.

KeyboardNerd replied:
Also, please rebase to master, and test. We found a bug in our tests that prevents database test failure from showing up.

Allda replied:
The version is usually included in CPE (in this case it is 7), but from the definition of CPE the version can be omitted.

https://access.redhat.com/blogs/766093/posts/2387501

Allda replied:
Yes, I noticed the bug. Today I spend a few hours trying to figure out why my test is not failing and after I found the cause and I fixed it I noticed that you fixed it yesterday

KeyboardNerd commented on database/pgsql/feature.go:
```diff
@@ -359,13 +360,14 @@ func (tx *pgSession) findFeatureIDs(fs []database.Feature) ([]sql.NullInt64, err
 	}
 
-	fMap := map[database.Feature]sql.NullInt64{}
+	fMap := map[string]sql.NullInt64{}
```
why do we need the string as key?

Allda replied:
That's probably forgotten code when potential namespace was part of Feature and not LayerFeature. I will fix it.

KeyboardNerd commented on database/pgsql/layer_test.go:
```diff
@@ -135,7 +157,8 @@ func TestPersistLayer(t *testing.T) {
 				assert.Nil(t, err)
 				assert.True(t, ok)
-				database.AssertLayerEqual(t, test.layer, &layer)
+				result := database.AssertLayerEqual(t, test.layer, &layer)
```
it already has the assertion inside.

KeyboardNerd reviewed (commented):
A few points, I'll do extra testing today.

Allda committed:
Use struct as a map key instead of string

String was used when Feature contains PotentialNamespace. Since it was
moved to LayerFeature we can use struct as map key instead of string.

KeyboardNerd reviewed (approved):
lgtm
